Tests/PackageLoading/ConventionTests.swift

```diff
@@ -94,3 +127,163 @@ private func fixture(files: [RelativePath], file: StaticString = #file, line: UI
+    var ignoreOtherModules: Bool = false
+
+    @discardableResult
+    init(_ path: AbsolutePath, in fs: FileSystem, package: PackageDescription.Package? = nil, _ body: @noescape (PackageBuilderTester) -> Void) {
+        do {
+            let loadedPackage = try loadPackage(path, in: fs, package: package)
+            result = .package(loadedPackage)
+            uncheckedModules = Set(loadedPackage.allModules)
+        } catch {
+            let errorStr = String(error)
+            result = .error(errorStr)
+            // FIXME: We need to extend dianostics to load warnings too.
+            // Right now warnings are dumped directly to stdout, we can use OutputByteStream
+            // instead to write warnings to a stream and then capture them for testing.
+            uncheckedDiagnostics.insert(errorStr)
+        }
+        body(self)
+        validateDiagnostics()
+        validateCheckedModules()
+    }
+
+    private func validateDiagnostics() {
+        guard !ignoreDiagnostics && !uncheckedDiagnostics.isEmpty else { return }
+        XCTFail("Unchecked diagnostics: \(uncheckedDiagnostics)")
+    }
+
+    private func validateCheckedModules() {
+        guard !ignoreOtherModules && !uncheckedModules.isEmpty else { return }
+        XCTFail("Unchecked modules: \(uncheckedModules)")
+    }
+
+    func checkError(_ str: String) {
+        switch result {
+        case .package(let package):
+            XCTFail("\(package) did not have any error")
+        case .error(let error):
+            uncheckedDiagnostics.remove(error)
+            XCTAssertEqual(error, str)
+        }
+    }
+
+    func checkModule(_ name: String, _ body: (@noescape (ModuleResult) -> Void)? = nil) {
+        guard case .package(let package) = result else {
+            return XCTFail("Expected package did not load \(self)")
+        }
+        guard let module = package.allModules.first(where: {$0.name == name}) else {
+            return XCTFail("Module: \(name) not found")
+        }
+        uncheckedModules.remove(module)
+        body?(ModuleResult(module))
+    }
+
+    final class ModuleResult {
+        private let module: Module
+        private lazy var sources: Set<RelativePath> = { Set(self.module.sources.relativePaths) }()
+
+        private init(_ module: Module) {
+            self.module = module
+        }
+
+        func check(c99name: String? = nil, type: ModuleType? = nil, isTest: Bool? = nil) {
+            if let c99name = c99name {
+                XCTAssertEqual(module.c99name, c99name)
+            }
+            if let type = type {
+                XCTAssertEqual(module.type, type)
+            }
+            if let isTest = isTest {
+                XCTAssertEqual(module.isTest, isTest)
+            }
+        }
+
+        func checkSources(root: AbsolutePath) {
+            XCTAssertEqual(module.sources.root, root)
+        }
+
+        func checkSources(_ paths: RelativePath...) {
```

Comment:
Should this be `checkSources(root:paths)`? If not maybe `checkSources(hasRoot:)` reads better?